### PR TITLE
feat(engine): warn on max_tokens_per_story at session boundaries

### LINKED ISSUE #336: max_tokens_per_story is checked only post-done and only journals — warn while the story runs
`limits.max_tokens_per_story` (default 2M, `policy.py:143`) is read in exactly one place, and that place runs *after* the story is already done. A story that defers or escalates — the outcome in the umbrella field report — is never checked at all. The reporter burned 8.35M weighted tokens against a 2M cap and learned about it from nothing.

## Evidence (validated at `3c68153`)

- **Sole read site:** `engine.py:1768-1775`, inside `_finalize_commit_phase`, *after* `advance(task, Phase.DONE)` (`engine.py:1764`). It appends one `token-budget-exceeded` journal line and does nothing else — no notify, no state.
- **Zero consumers** of that journal kind anywhere. The TUI and `status` show totals only, never budget or headroom (`cli.py:1567-1580`, `tui/widgets.py:165-177`).
- **Worse than "late":** the check lives only on the commit path. A story that defers or escalates gets no entry whatsoever. Total silence, not late silence.
- **#158 does not cover this.** The per-session guard (`max_tokens_per_session`, default 4M, `adapters/generic.py:605-683`) is a different cap, latched per session, and blind to cross-session story accumulation. Nothing budget-related has landed since 0.9.0.
- `core.toml:144-148` is missing `label` and `description` — the only `limits` field without them.

## Fix outline

- **No new knob** — the cap is advisory by contract (`docs/FEATURES.md:145`); warn at 100%.
- Delete the post-done check (`engine.py:1768-1775`). Add `_note_story_token_budget(task)` called from `_run_session` right after `self._save()` (`engine.py:2709`), where `task.tokens` is both story-cumulative and fresh. Sweep and Stories engines inherit it for free.
- Would have fired after session 1–2 of the reported run.
- Reuse the `token-budget-exceeded` journal kind (zero `src/` consumers) and add a `budget` field. `gates.notify` title `story token budget exceeded: <key>`, message carrying weighted / cap / raw plus "advisory: story continues".
- Persisted latch `StoryTask.token_budget_warned: bool = False` — resume must not re-notify.
- Weight comes from live policy `limits.cache_read_weight` (enforcement side of the doctrine comment at `engine.py:615-623`; display uses the persisted `RunState.cache_read_weight()` snapshot, `model.py:399`).
- `core.toml:144-148`: add the missing `label` + `description` (presentation-only, no sync-test impact).
- Docs: reword `FEATURES.md:145` and `README.md:593`.
- Zero changes to the #158 per-session guard — `SessionSpec`, `generic.py`, `opencode_http.py` untouched.

## Explicit non-goals (follow-ups)

- **Adapter-level mid-session story check** (~30s granularity): would add `story_token_budget` / `story_tokens_spent` to `SessionSpec` (`adapters/base.py:75-78`), populate at `engine.py:2546-2549`, and extend the trip condition in `generic.py:608-614` + `opencode_http.py:1067-1073` — needs a latch separate from #158's single `budget_tripped` boolean.
- **Budget headroom display** in `status` and the TUI (free, independent of the above).

## Tests

Amend `test_engine.py:1235-1264` (still exactly one journal entry, weighted = 160000, plus ATTENTION, the `budget` field, and the latch); keep `:1065-1098` untouched as the no-false-positive case; new warns-on-a-deferring-story test (ablation: re-guard with `phase == DONE` → it must fail); warns-once test (ablate the latch); latch-survives-resume test (ablate `to_dict`); `test_model.py` round-trip. Settings sync tests must pass unchanged.

Edge cases already analysed: resume-replay sessions don't hit the new site (the persisted latch covers it); aborted sessions skip judgment; sweep bundles inherit the behaviour.

Part of the 0.9.0 field-report umbrella (linked parent).

### PR DESCRIPTION
**Pull request by pbean**
`max_tokens_per_story` (2M weighted, advisory) had exactly one read site: inside
`_finalize_commit_phase`, *after* `advance(task, Phase.DONE)`. Two consequences,
both hit by the 0.9.0 field report behind #337:

- the overrun was reported only once every token had already been spent, and
- a story that **defers or escalates** never reached that line at all — the
  reported run burned 8.35M weighted against the 2M cap and produced no record
  of any kind. Total silence, not late silence.

## What changed

The story's cumulative weighted spend is re-checked at **every session
boundary**: a new `_note_story_token_budget(task)` called from `_run_session`
right after its tail `_save()`. That is the point where `task.tokens` is both
fresh (`attach_session_usage` folded this session in just above) and
story-cumulative, and every session of every run type funnels through it — so
deferring and escalating stories are judged like committing ones, and
`SweepEngine` / `StoriesEngine` inherit it unchanged. A session that aborted
never reaches the site (the exception propagates past the `finally`), so no
story is judged on a usage read that did not happen.

The first crossing:

- journals the existing `token-budget-exceeded` kind, now carrying `budget`
  alongside `weighted` / `total`;
- raises an ATTENTION + desktop notice — `story token budget exceeded: <key>`,
  body naming both units, the cap, and that this is advisory and the story
  continues.

It fires **once per story**, latched on a new persisted
`StoryTask.token_budget_warned` — every session after the crossing is over the
cap too, and the latch is in `to_dict`/`from_dict` precisely so a resumed run
does not re-notify what the pre-pause process already did.

Enforcement weight is the live `self.policy.limits.cache_read_weight`, the
enforcement side of the split documented at `summary()` (displays must
reproduce from `state.json`; budgets bind at the policy the running process
enforces).

Still advisory — nothing is terminated. **No new policy knob.** Zero changes to
the #158 per-session guard: `SessionSpec`, `adapters/generic.py` and
`adapters/opencode_http.py` are untouched.

Also: `core.toml`'s `max_tokens_per_story` entry was the only `[limits]` field
with no `label`/`description` (presentation-only; the settings sync tests are
unchanged and pass as-is).

## Tests

- `test_token_budget_exceeded_journals_weighted` amended — still exactly one
  entry at `weighted=160000`, plus the `budget` field, the ATTENTION content and
  the latch.
- `test_token_budget_discounts_cache_reads` untouched — the no-false-positive
  case still holds through the new site.
- new `test_token_budget_warns_on_a_story_that_never_commits` (the #336 case),
  `test_token_budget_warns_once_per_story`, `test_token_budget_latch_survives_resume`.
- `test_model.py` round-trip + legacy-state default.

Per the repo's ablation rule, each negative assertion was proven by deleting its
gate first:

| Ablation | Test that must fail | Result |
| --- | --- | --- |
| re-guard the check with `task.phase != Phase.DONE: return` | `..._warns_on_a_story_that_never_commits` | fails (4 of 5 fail — the placement is load-bearing) |
| delete the `token_budget_warned` early return | `..._warns_once_per_story` | fails (2 entries) |
| drop `token_budget_warned` from `to_dict` | `..._latch_survives_resume` | fails; the runtime-latch test still passes, so the two gate different things |
| move `token_budget_warned = True` back above `journal.append` | `..._latch_waits_for_the_durable_record` | fails (latch `True` on disk, record absent) — and only this test |
| delete the `except OSError` around the ATTENTION write | `test_notify_file_swallows_an_unwritable_attention_path` | fails (`IsADirectoryError` escapes) |
| delete the `max_tokens_per_story` bool/int guard | `..._rejects_non_integers` | fails all 3 cases |
| delete the `max_tokens_per_story >= 1` guard | `..._must_be_positive` | fails both cases |

`uv run pytest -q -n auto`: 3416 passed, 24 skipped. Full `trunk check` clean;
`pyright@1.1.411` (CI pin) 0 errors.

## Review round 1 — two bots, one finding, plus two defects they surfaced

Both reviewers flagged the same thing: the latch `task.token_budget_warned = True` was only made
durable by the trailing `self._save()`, after `journal.append` and `gates.notify`. Both proposed
hoisting the `_save()` up.

**Declined the hoist, fixed the actual defect at the other end of the block** (2771189). The
window they describe is narrower than stated — `_run_inner`'s `finally` at `engine.py:491` already
`_save()`s on every abort arm, so only SIGKILL reaches it, and a duplicate breaks nothing
downstream. The hoist would convert at-least-once into at-most-once: a death between the save and
the append leaves the latch durable with no record and no notice, i.e. the #336 silent overrun
reintroduced.

The real bug was one line *above*: the latch was set **before** the `journal.append`. That append
is unguarded IO (`journal.py:41-42`), so an `OSError` let the same `finally` persist the latch with
nothing written anywhere — permanent silence, no SIGKILL needed. Unlike the phase `_record_defer`
and `_escalate` set before their own writes, this bit is rendered on no surface, so a lost record
is unrecoverable. The latch now sits between the append and the notify; the `journal → notify →
save` emit order is unchanged, so it still matches every other record-a-decision site.
`test_token_budget_latch_waits_for_the_durable_record` fails on 9a15718.

Two further defects surfaced during that review and are fixed here too (maintainer's call to
handle in-PR rather than as follow-ups) — **this makes the PR three logical changes rather than
one, kept as three separate commits**:

- **5cbb7d1 — `gates.notify` didn't honor its "never raises" contract.** The ATTENTION append
  (`gates.py:94`) was bare IO while only the desktop half was guarded, so an unwritable run dir
  turned an advisory notice into a run crash at all ~11 record-a-decision sites — several of which
  notify *after* journaling a decision they cannot unwind. Guarded inside `notify` so every call
  site is covered at once, matching the observe-degrade rule already applied to this same call at
  `adapters/generic.py:633-637`.
- **fcc510f — `limits.max_tokens_per_story` had no validation.** Bare `int()` at `policy.py:743`,
  so `true` became a 1-token cap and `0` was accepted, against the documented `int >= 1`
  (`core.toml` already sets `minimum = 1`, so the settings TUI could never produce either).
  Mirrors `max_tokens_per_session`'s bool/int + `>= 1` guards. ⚠️ **Compat flag for the
  maintainer:** this rejects a `policy.toml` that previously loaded. There is no `0 = off`
  semantic — set the cap high instead, it only warns.

## Explicit non-goals (follow-ups)

Both were considered and deliberately left out of this PR:

1. **Adapter-level mid-session story check** — plumbing the story budget into
   `SessionSpec` so the running session itself samples against it (~30s cadence,
   like the #158 per-session guard). Needs a latch separate from #158's single
   `budget_tripped` boolean. This PR's session-boundary check would already have
   fired after session 1–2 of the reported run.
2. **Budget headroom in `status` / TUI** — both surface totals only, with no
   budget or remaining-headroom column (`cli.py` status text, `tui/widgets.py`
   token columns).

Closes #336

## Summary by CodeRabbit

* **Changed**
  * Story token budgets are now checked after every session boundary, including deferred or escalated sessions.
  * Exceeding the advisory cap shows an attention and desktop notification while the story continues running.
  * Budget warnings are recorded once per story and remain suppressed after resuming.
  * Token-budget journal entries now include the configured budget.
  * Updated settings and documentation clarify weighted token accounting and the distinction between story and session budgets.

### CODE REVIEW

File: src/bmad_loop/engine.py [hunk 1]
```diff
@@ -2762,6 +2755,59 @@ def _run_session(
...
+        gates.notify(
+            self.policy,
+            self.run_dir,
+            f"story token budget exceeded: {task.story_key}",
+            f"{weighted} weighted tokens ({task.tokens.total} raw incl. cache reads) "
+            f"past the {budget} `limits.max_tokens_per_story` cap — advisory: the "
+            f"story continues. Stop the run with `bmad-loop stop` if the spend is "
+            f"no longer worth it.",
+        )
+        self._save()
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/engine.py]: Good catch on the direction — there **is** an ordering defect in this block. But it's at the other end, and hoisting `_save()` would make things worse rather than better. Fixed in fcc510f (well, 2771189 for this part).

## Why not the `_save()` hoist

**The stated window is much narrower than it looks.** `_run_inner` has a `finally` at `engine.py:491` that calls `self._save()` unconditionally on *every* arm — clean finish, `RunPaused`, `RunStopped` (graceful and hard), `KeyboardInterrupt`, and the generic `except Exception` at 441-481. `engine.py:451` says so out loud: *"Reset before the nested re-raise so the trailing `_save()` persists it on both paths."* So every in-process abort — exception, SIGTERM/SIGINT, `bmad-loop stop`, plugin fault — already persists the latch. Only a SIGKILL/power-loss reaches the window, and its cost is one duplicate ATTENTION line plus one duplicate journal entry. I checked the consumers: the only aggregator over `token-budget-exceeded` is `diagnostics.py:378-389`'s descriptive `kind_histogram` (the derived metrics at 407-409 are escalation/defer/plugin-error only); the TUI reads the journal as a log tail and ATTENTION as a text blob. A duplicate breaks nothing.

**The hoist trades that for a strictly worse failure.** Under `save → journal → notify`, a death between the save and the append leaves the latch durable with **no journal entry and no ATTENTION line** — the overrun becomes permanently invisible. That is exactly the #336 bug class this PR exists to remove: it converts at-least-once into at-most-once, on a signal whose entire purpose is to not be silent.

**The current emit order is also the uniform house pattern**, not an oversight — `_record_defer` (`engine.py:3135/3141/3147`), `_escalate` (3242/3243/3249), `_vetoed` (998-1014), `recovery_flow.py:635/641/647`, `worktree_flow.escalate_unit` (651/652/658). No site in the tree saves before notifying the event it is announcing.

## The real defect, at the top of the block

`task.token_budget_warned = True` was set at line 2792, **before** the `journal.append` at 2793. `Journal.append` is unguarded IO (`journal.py:41-42`, bare `open("a")` + `write`), and there's no intermediate catch between the call site (`engine.py:2745`) and the run-level crash arm. So if that append raises `OSError`, the `finally` at 491 persists `token_budget_warned: true` with **no journal entry and no notification**, and the check early-returns at 2786 forever after. Total permanent silence on an overrunning story, from an ordinary IO fault — no SIGKILL required.

What makes this site different from its siblings: at `_record_defer`/`_escalate` the pre-set mutation is `Phase.DEFERRED`/`Phase.ESCALATED`, visible in `status`, `--json`, the TUI and `summary()` — a lost journal write still leaves the decision legible. `token_budget_warned` is rendered **nowhere** (it appears only in `model.py` serialization, `engine.py:2786/2792`, and tests). It is a pure write-only suppression bit whose only job is to suppress the one artifact carrying the information.

**Fix: move the latch down one position**, so it sits between the journal write and `gates.notify`. The `journal → notify → save` emit order is byte-identical; only the latch moves behind the durable record. Invariant now recorded in the code at both ends so neither gets "fixed" again: *never latch a suppression bit before the record it suppresses is durable*, and `_save()` stays last because at-least-once is the right bias here.

## The regression test you asked for

`test_token_budget_latch_waits_for_the_durable_record` (`tests/test_engine.py`) — patches `journal.append` to raise `OSError` for the budget kind, asserts the run crashed **and** that the persisted latch is `False`, then resumes and asserts the owed warning is actually delivered. **It fails on the previous commit** (latch `True` on disk, resume silent). Ablation run: putting the latch back above the append fails this test and only this test.

One test I deliberately did *not* write: raising from `gates.notify` can't distinguish the two orderings — under both, the `finally` at 491 persists the latch and one entry exists. It would pin untouched code rather than the ordering.

## Two more defects this surfaced (also fixed here, at the maintainer's direction)

- **`gates.notify` didn't honor its own "never raises" contract** — the ATTENTION append at `gates.py:94` was bare IO while only the desktop half was guarded, so an unwritable run dir turned an advisory notice into a run crash at all ~11 record-a-decision sites. Guarded inside `notify` so every call site is covered at once, matching the observe-degrade rule already applied to this same call at `adapters/generic.py:633-637`. (5cbb7d1)
- **`limits.max_tokens_per_story` had no validation** — bare `int()` at `policy.py:743`, so `true` became a 1-token cap and `0` was accepted, against the documented `int >= 1`. Mirrored `max_tokens_per_session`'s bool/int and `>= 1` guards. Flagging that this rejects a `policy.toml` that previously loaded; there's no `0 = off` semantic. (fcc510f)

## On the docstring-coverage pre-merge check

Dismissing that one as tool-config noise: `pyproject.toml:79` selects only `["E4", "E7", "E9", "F"]` — no `D`/pydocstyle — and there's no interrogate/darglint in `.trunk/trunk.yaml` or CI. `AGENTS.md` explicitly says not to ratchet ruff stricter than CI. `tests/test_model.py` has 41 functions and 0 docstrings as its established convention, which the two tests added there match; the new production method carries a 25-line docstring.

Full suite 3416 passed / 24 skipped, `trunk check` clean, `pyright@1.1.411` (the CI pin) 0 errors.

[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/engine.py]: Same finding as the CodeRabbit thread — full reasoning there: https://github.com/bmad-code-org/bmad-loop/pull/346#discussion_r3663321204

Short version: the direction is right but the `_save()` hoist is the wrong end of the block.

- `_run_inner`'s `finally` at `engine.py:491` already `_save()`s on **every** abort arm (including the generic `except Exception` at 441-481 — see the comment at 451), so only a SIGKILL reaches the window you describe, and a duplicate breaks nothing downstream (`diagnostics.py:378-389`'s histogram is descriptive; the derived counters at 407-409 don't include this kind).
- The hoist converts at-least-once into at-most-once: a death between `_save()` and `journal.append` would leave the latch durable with no record and no notice — the #336 silent-overrun bug, reintroduced.
- `journal → notify → save` is also the uniform house order (`_record_defer` 3135/3141/3147, `_escalate` 3242/3243/3249, `_vetoed` 998-1014, `recovery_flow.py:635/641/647`).

**The actual defect was one line above**: `task.token_budget_warned = True` sat *before* the `journal.append`. `Journal.append` is unguarded IO (`journal.py:41-42`), so an `OSError` there let that same `finally` persist the latch with nothing written anywhere — permanent silence, no SIGKILL needed. Unlike the phase its siblings set pre-write, this bit is rendered on no surface, so a lost record is unrecoverable. Moved the latch below the append (emit order unchanged) in 2771189, with `test_token_budget_latch_waits_for_the_durable_record` — which fails on 9a15718.

Your point about the existing resume test not covering a crash *inside* the method was exactly right, and that's the gap the new test closes.
